### PR TITLE
Places uniffication

### PR DESCRIPTION
**Pull request by skhamis**
Companion to the PR in a-s: https://github.com/mozilla/application-services/pull/4689

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/PlacesStorage.kt [hunk 1]
```diff
@@ -130,10 +125,6 @@ abstract class PlacesStorage(
             logger.debug("Successfully synced.")
             SyncStatus.Ok
 
-            // Order of these catches matters: InternalPanic extends PlacesException.
-        } catch (e: InternalPanic) {
```
Comment (mhammond) [inline components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/PlacesStorage.kt:134]: IIUC, there's an "InternalException" in the uniffi generated kotlin which serves this purpose, which I guess we need to use? (There's another example above, but because it just rethrows without taking any other action, I think it's fine to omit it there?)

File: components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/Types.kt [hunk 2]
```diff
@@ -77,12 +77,39 @@ internal fun mozilla.appservices.places.VisitType.into() = when (this) {
...
+    VisitType.BOOKMARK -> mozilla.appservices.places.uniffi.VisitTransition.BOOKMARK
+    VisitType.EMBED -> mozilla.appservices.places.uniffi.VisitTransition.EMBED
+    VisitType.REDIRECT_PERMANENT -> mozilla.appservices.places.uniffi.VisitTransition.REDIRECT_PERMANENT
+    VisitType.REDIRECT_TEMPORARY -> mozilla.appservices.places.uniffi.VisitTransition.REDIRECT_TEMPORARY
+    VisitType.DOWNLOAD -> mozilla.appservices.places.uniffi.VisitTransition.DOWNLOAD
+    VisitType.FRAMED_LINK -> mozilla.appservices.places.uniffi.VisitTransition.FRAMED_LINK
+}
+
+internal fun mozilla.appservices.places.uniffi.VisitTransition.into() = when (this) {
+    mozilla.appservices.places.uniffi.VisitTransition.LINK -> VisitType.LINK
```
Comment (mhammond) [inline components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/Types.kt:94]: if we are keeping NOT_A_VISIT, don't we want to handle null here? Or maybe not?

File: components/browser/storage-sync/src/test/java/mozilla/components/browser/storage/sync/PlacesHistoryStorageTest.kt [hunk 3]
```diff
@@ -247,7 +246,7 @@ class PlacesHistoryStorageTest {
         assertEquals(VisitType.RELOAD, visits[1].visitType)
 
         assertEquals("http://www.firefox.com/", visits[2].url)
-        assertEquals("", visits[2].title)
+        assertEquals(null, visits[2].title)
```
Comment (mhammond) [inline components/browser/storage-sync/src/test/java/mozilla/components/browser/storage/sync/PlacesHistoryStorageTest.kt:249]: huh - this change is great, because if we see an incoming observation with an empty string as the title then we will actually set the title to that empty string. That fact that the old code used to turn a null into an empty string when querying might have been a footgun

File: components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/FxaDeviceSettingsCache.kt [hunk 4]
```diff
@@ -33,15 +33,15 @@ class FxaDeviceSettingsCache(context: Context) : SharedPreferencesCache<DeviceSe
         return JSONObject().also {
             it.put(KEY_FXA_DEVICE_ID, this.fxaDeviceId)
             it.put(KEY_DEVICE_NAME, this.name)
-            it.put(KEY_DEVICE_TYPE, this.type.name)
+            it.put(KEY_DEVICE_TYPE, this.kind.toString())
```
Comment (mhammond) [inline components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/FxaDeviceSettingsCache.kt:36]: hrm, these are unrelated and due to other pending stuff, right? (the sync-manager stuff sure looks like it, but I don't recall fxa changes.)

**Review by mhammond — COMMENTED** [review]
That NOT_A_VISIT thing is a bit annoying, but this looks great to me. I see nothing at all related to the timestamp, which surprises me, but also makes me happy :)

File: components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/FxaDeviceSettingsCache.kt (hunk 4 above)
Comment (skhamis) [reply components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/FxaDeviceSettingsCache.kt:36]: Yeah I tried to split the commits to shove as much of the sync manager changes as possible into one. There were indeed some changes to FxA though from Ben's PR https://github.com/mozilla-mobile/android-components/pull/11256/files#diff-8fff056ceecd8901d1e9ed455b4e23eac1232c765bbbc06aa0151991e7c9f37c

File: components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/PlacesStorage.kt (hunk 1 above)
Comment (skhamis) [reply components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/PlacesStorage.kt:134]: yes! We package InternalException under PlacesException now instead as it's own hierarchical error on the android side. Semantically they were doing the same things IIUC

**Comment by skhamis** [comment]
> That NOT_A_VISIT thing is a bit annoying, but this looks great to me. I see nothing at all related to the timestamp, which surprises me, but also makes me happy :)

I updated the other PR to not use unsigned and just handle the conversion in the UniffiWrapperType so that this could be much cleaner and feel it definitely works well!
